cachito/workers/pkg_managers/golang.py

```diff
@@ -123,10 +131,150 @@ def resolve_gomod_deps(app_source_path, request_id, dep_replacements=None):
+        return f'v{module_major_version or "0"}.0.0-{commit_timestamp}-{commit_hash}'
+
+    # An example of a semantic version with a prerelease is v2.2.0-alpha
+    if semantic_version.prerelease:
+        # vX.Y.Z-pre.0.yyyymmddhhmmss-abcdefabcdef is used when the most recent versioned commit
+        # before the target commit is vX.Y.Z-pre
+        version_seperator = '.'
+        pseudo_semantic_version = semantic_version
+    else:
+        # vX.Y.(Z+1)-0.yyyymmddhhmmss-abcdefabcdef is used when the most recent versioned commit
+        # before the target commit is vX.Y.Z
+        version_seperator = '-'
+        pseudo_semantic_version = semver.bump_patch(str(semantic_version))
+
+    return f'v{pseudo_semantic_version}{version_seperator}0.{commit_timestamp}-{commit_hash}'
+
+
+def get_golang_version(module_name, git_path, target_ref, update_tags=False):
+    """
+    Get the version of the Go module in the input Git repository in the same format as `go list`.
+
+    If target_ref doesn't point to a commit with a semantically versioned tag, a pseudo-version
+    will be returned.
+
+    :param str module_name: the Go module's name
+    :param str git_path: the path to the Git repository
+    :param str target_ref: the Git reference of the Go module to get the version for
+    :param bool update_tags: determines if `git fetch --tags --force` should be run before
+        determining the version
+    :return: a version as `go list` would provide
+    :rtype: str
+    """
+    # If the module is version v2 or higher, the major version of the module is included as /vN at
+    # the end of the module path. If the module is version v0 or v1, the major version is omitted
+    # from the module path.
+    module_major_version = None
+    match = re.match(r'(?:.+/v)(?P<major_version>\d+)$', module_name)
+    if match:
+        module_major_version = int(match.groupdict()['major_version'])
+
+    repo = git.Repo(git_path)
+    if update_tags:
+        repo.remote().fetch(force=True, tags=True)
+
+    commit_to_tags_info = {}
+    not_semver_tag_msg = '%s is not a semantic version tag'
+    for tag in repo.tags:
+        if not tag.name.startswith('v'):
+            log.debug(not_semver_tag_msg, tag.name)
+            continue
+
+        try:
+            # Exclude the 'v' prefix since this is required by Go, but it is seen as invalid by
+            # the semver Python package
+            parsed_version = semver.parse_version_info(tag.name[1:])
+        except ValueError:
+            log.debug(not_semver_tag_msg, tag.name)
+            continue
+
+        commit_to_tags_info.setdefault(tag.commit.hexsha, []).append({
+            'parsed_version': parsed_version,
+            'tag_name': tag.name,
+        })
+
+    target_commit = None
+    for commit in repo.iter_commits(target_ref):
+        if not target_commit:
+            target_commit = commit
+
+        for tag_info in commit_to_tags_info.get(commit.hexsha, []):
```

Comment:
What if there is a commit tagged with multiple semantic versions? I think it may just arbitrarily pick one of them? Should we consider this an error, pick the highest one, or something else?

Reply:
I've been trying to figure out what algorithm Go uses to pick the right tag when there are multiple tags on the same commit. I'm getting results that differ from what I've read in the code. I made a post and will try again on Monday.
https://forum.golangbridge.org/t/how-is-a-version-determined-on-a-commit-with-multiple-tags/16265